### DISCUSSION TITLE
Reduce vertical space by going borderless

### DISCUSSION THREAD

**Discussion by coppolaemilio** [godotengine/godot-proposals#4648] [Editor]
### The problem

Currently the editor has an top bar with a lot of empty space. I believe this elements can be moved somewhere else to better use the vertical space while in editor.

![empty](https://user-images.githubusercontent.com/2206700/172910578-581747f9-f40c-4ef4-9cc2-ca29157ac10c.png)

-----
### An existing proposal

There is this proposal https://github.com/godotengine/godot-proposals/issues/492 which tries to merge the tabs and the menus in one line:

![empty](https://user-images.githubusercontent.com/42384785/74615261-2bc4a400-50ed-11ea-956c-8dca5347a379.jpg)

Unfortunately, this only works in the default layout, as soon as you start moving the panels around you will find many issues when implementing this solution. A user highlighted this issue showing their preferred layout:

![image](https://user-images.githubusercontent.com/2206700/172912652-e1862803-48cb-4e51-a5cd-a5806d48c3ee.png)

So, if we have to keep that line of content in a separated row we can't take the space from the other elements already existing in the editor.

-----

### A new proposal

Instead of merging those two, the obvious solution would be to get rid of the "border" of the window and go with an integrated border-less mode. Applications like Figma, Spotify, Firefox or VSCode already do this:
![Screen Shot 2022-06-09 at 19 43 47](https://user-images.githubusercontent.com/2206700/172910994-582cefdc-e907-45b5-badd-e6f803ede05f.png)

Integrating the system minimize, maximize or close with the menu and run options could save vertical space.

This could look something like this:
![Screen Shot 2022-06-09 at 20 01 45](https://user-images.githubusercontent.com/2206700/172914201-c33a9acf-c5e5-42a0-a89f-305c99d34b71.png)

(side by side)
![image](https://user-images.githubusercontent.com/2206700/172914330-37eec681-0ae0-4f8d-8ff6-debad831a003.png)

Further more, in systems like Mac or some Linux distributions, you can take advantage of the native menu bar and move the options there. This should be already be possible thanks to https://github.com/godotengine/godot/pull/59410

![image](https://user-images.githubusercontent.com/2206700/172914840-1bddd18e-ee5b-4afa-be23-22b7a8cef384.png)

Don't look too close to the examples and mockups I made since they are not pixel perfect, just a demonstration.

This borderless style should be optional since there are many desktop environments that won't be able to deal with this kind of layout. There should also be options to choose where to place the close/minimize/maximize options (left or right).

Let me know if you have any questions or if you have any issues with this approach, I'm happy to keep updating the discussion answering any question that you have.

(Of course, after publishing and searching again I found another related discussion: https://github.com/godotengine/godot-proposals/discussions/4132 adding it as a reference)

**Comment by graugraugrau** [comment]
In general I like this idea, but it may not work well on Linux. There are a lot of desktop environments (e.g. Gnome, KDE, Xfce, i3, ...) that on top also allow configuration on the window titlebar (placement of close at the right/left, hide minimize, hide maximize). This would lead to some further design topics and more OS dependencies.
I'm not saying it's impossible, but it's probably not that simple for Linux and the different desktop environments.

**Reply by Calinou** [reply]
Client-side decorations are likely the way to go if you want to save screen real estate. I know this won't look as consistent with the rest of the OS, but I feel the saved screen real estate alone is an usability upgrade.

To let users best match their desktop environment, it's possible to provide an editor setting to control the close button position (defaulting to left on macOS and right on other platforms).

**Reply by coppolaemilio** [reply]
@griseus I see your point. I made a mac mockup because that's the only OS I have available at the moment, but maybe this update can also be "optional" and leave the regular borders as a fallback in case the OS doesn't support this type of windows. 

I like what Calinou is saying about allowing to customize the location of the close/minimize/maximize buttons, so it might be doable?

**Reply by graugraugrau** [reply]
I would probably deactivate it. Other cross platform IDEs seem to not solve this problem either. For example Visual Code, Rider, MonoDevelop or MATLAB.

**Reply by coppolaemilio** [reply]
Well, it seems like it should be an optional setting then, but I wonder if there is any project that deals with this gracefully 🤔 
(I updated the description to clarify that it can be just an option, not the only way to see the main editor)

**Reply by Calinou** [reply]
Regular borders will have to remain supported no matter what, as mobile platforms and HTML5 don't allow using client-side decorations – there is no global menu to speak of there. (The Godot editor can run on mobile platforms and HTML5.)

**Reply by eon-s** [reply]
And don't forget that addons can add more items there too (in menu, toolbar, etc.), this will make it harder to manage...

**Comment by sztrovacsek92** [comment]
Good idea, but it might be a bit overengineered (I'm also not a big fan of custom windows :D).

Wouldn't this be the easiest solution?
![godot](https://user-images.githubusercontent.com/67164514/172945610-f0be99a2-9243-4c61-aa02-8183007d4ff0.jpg)

Just ditch the horizontal tabs for the scenes, and use the already existing script selector (which can conveniently be hidden, by a click of a button at the bottom), to browse between scenes. And we can put the whole top shelf in the newly created space.

**Reply by coppolaemilio** [reply]
Borderless mode would be optional. Your proposal is already discussed here: https://github.com/godotengine/godot-proposals/issues/492

Don't want to get too deep discussing your version but, that might work in a very wide screen and with the default layout, but as soon as you start opening new tabs, move panels to different places or your screen is smaller and you have more stuff open it won't work.

**Reply by sztrovacsek92** [reply]
I've seen #492, but as far as I can tell, they only discuss putting the top shelf above the sidebars.
Yeah, it's almost the same, but still different. In this case the shelf will be always on top of the Viewport, doesn't matter if you only have 1 sidebar or 4.

I don't think there's a need for a very wide screen, but we can still group the Play/Stop/etc buttons (with responsiveness, to only group on smaller sizes), we could also get rid of the GLES selector button and put it in the Project menu (I never used it before), and we can even move some buttons one shelf down, next to the help links.

**Reply by coppolaemilio** [reply]
There is still no place for the tabs 🤔

**Reply by sztrovacsek92** [reply]
Which tabs? The scene selector is already there.

**Reply by Calinou** [reply]
> we could also get rid of the GLES selector button and put it in the Project menu (I never used it before),

One reason the rendering backend is always present on the editor is to make it visible on screenshots. As an engine contributor, this is useful to diagnose issues when someone posts a screenshot of the editor. (People don't always post their project files whenever they post a screenshot of the editor.)

This also applies to the Godot version number at the bottom of the window.

**Reply by sztrovacsek92** [reply]
@Calinou than just put that button next to the version number, maybe?
(Does it even need to be a button? Plain text should be enough.)

**Reply by coppolaemilio** [reply]
But the button has the functionality to change the rendering backend, so making it plain text won't look like something you can actually change

**Reply by Calinou** [reply]
> @Calinou than just put that button next to the version number, maybe? (Does it even need to be a button? Plain text should be enough.)

Having it as plain text should be OK, but the backend + version label can't be made too long. Horizontal space in the bottom panel is scarce on 1366×768 displays, which are still prevalent in budget laptops used by many Godot users.

**Reply by sztrovacsek92** [reply]
> But the button has the functionality to change the rendering backend, so making it plain text won't look like something you can actually change

True, but how often do you actually need to change the renderer? Imho it's a more advanced feature, not something that should be one click away for every new user.

> Horizontal space in the bottom panel is scarce

No need to go fancy there, something like _3.4.4.s ES3_ (or VK for Vulkan), should be more than enough - and it's even 1 letter shorter than the current _3.4.4.stable_.

> 1366×768 displays, which are still prevalent in budget laptops used by many Godot users.

I'm one of them, I know. :D

**Reply by coppolaemilio** [reply]
> Which tabs? The scene selector is already there.

These are the tabs I mean:
<img width="1512" alt="Screen Shot 2022-06-10 at 01 15 24" src="https://user-images.githubusercontent.com/2206700/172960419-0f3e48ff-0db1-4308-84bc-ba179ad5a88a.png">

Edit: I see that you did the side bar to the main panel, but that would be take much more relevant realstate than the tabs ever will on the main viewport

**Reply by sztrovacsek92** [reply]
You can show/hide it with one click, there's a button at the bottom.

**Reply by coppolaemilio** [reply]
This is starting to sound more over-engineered than making the window borderless 😅 
I could even argue that even your new layout could be better being borderless by just moving the window controls to that same line.

**Reply by sztrovacsek92** [reply]
But... but... I'm only using built-in Control nodes, duh! :smile:

**Reply by Jake012345** [reply]
Doesn't matter in basic... From now on, godot-android has the high-ground 😎

**Comment by azur-wolve** [comment]
Godot 4 will have multi-screen support, to take into account.

**Reply by Calinou** [reply]
Godot 4.0's editor will not have full multi-screen support. It'll only have splitting docks to separate windows and using separate windows for dialogs (such as popups, project/editor settings, etc).

Full multi-screen support is planned for a future 4.x release.

**Reply by azur-wolve** [reply]
it will allow splitting also the workspaces?

**Reply by Calinou** [reply]
No, not in 4.0. This is also planned for a future release.

**Comment by Goutte** [comment]
On **KDE** I always use `ALT+F3, M, N` to remove the window borders in Godot (and Firefox).   Not as tight as the proposed examples, but it already provides some comfort.

**Comment by LimaneGaya** [comment]
What's the state of this proposal? i would really like to get rid of the title bar on linux as it takes unnecessary space.

**Reply by Calinou** [reply]
See https://github.com/godotengine/godot-proposals/discussions/4648#discussioncomment-12511104.

**Comment by Calinou** [comment]
Closing in favor of https://github.com/godotengine/godot-proposals/issues/4846. This is now implemented on macOS; support for other platforms requires changes to the platform code to make client-side decorations possible.
